### PR TITLE
feat(opsem): track alloc/free for memory objects

### PR DESCRIPTION
**Pull request by priyasiddharth**
1. Generalize Tracking memory managers to track read, writes and allocations.

2. Add builtins for
isRead(...)
resetRead(...)
isAlloc(...)
free(...)

3. read instructions are currently not wired in opsem

### CODE REVIEW

**Comment by priyasiddharth** [comment]
Looking at test failures...

**Comment by priyasiddharth** [comment]
Ready for review...

File: lib/Analysis/SeaBuiltinsInfo.cc
Comment (agurfinkel) [inline lib/Analysis/SeaBuiltinsInfo.cc:347]: should also be marked with `doesNotFreeMemory()`. While you can the function `free`, a more appropriate name is `mark_free`. The function marks a memory region as freed, but it does not free the memory region itself. The freeing will be done by calling libc `free` or C++ `delete`.

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc
Comment (agurfinkel) [inline lib/Transforms/Scalar/PromoteVerifierCalls.cc:219]: the bodies of these if statements are all the same, why not refactor the code into a (lambda) function and set a map between function name and function that replaces it. So the code looks like that:
```
if (map.count(fn->getName())) {
   replace(fn, map.at(fn->getName());
 }
```

File: lib/seahorn/BvOpSem2Context.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2Context.hh:20]: use `enum class`

File: lib/seahorn/BvOpSem2Context.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2Context.hh:710]: comment does not match method name

File: lib/seahorn/BvOpSem2FatMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2FatMemMgr.cc:561]: is this correct? I am not sure what `setMemory` supposed to do, but it is strange that all slots are rewritten using the same value

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:798]: should be `ERR` to be consistent with other usage

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:821]: there should be a warning printed if assertion fails. This can be hard to trace in non-debug mode

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:820]: the name `setMemory` is confusing because it is interpreted either as a setter (which it is not), or something that changes memory region (which it is not). The implementation seems to be a wrapper over `FilledMemory` that returns initialized memory region. Maybe rename all into `InitMem` or something similar to reflect the intention of the function.

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:6]: why are all of these globals? Can we merge them into class definition?

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:149]: still a TODO?

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:151]: I am confused about this initialization. This should only be used when `store` operation fails. Should it be just a null indicating that operation failed?

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:141]: this function should be factored out into the mixin. It simply dispatches into appropriate store function based on type of value.

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:277]: should not the load be assumed to be aligned?

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:301]: interesting choice for initial value for havoced memory...

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:389]: all `not implemented` messages should go into `ERR`

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:54]: can use stack data structure like `llvm::ArrayRef` or `std::array`

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:105]: fragile. order of metadata and order of enum can be different. at least comment somewhere that they should match. But better, add a function that converts kind to index in tracking memory manager and use that function.

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:285]: I know believe that this should be `initializedMemory` and `zeroedMemory` should be `zeroInitializedMemory`
Note that `zeroinitialized` is a keyword LLVM uses to mark a global memory region `zeroinitialized`, so the name will make sense to anyone familiar with llvm

File: test/opsem2/trackingmem/freed_unsat.c
Comment (agurfinkel) [inline test/opsem2/trackingmem/freed_unsat.c:18]: I missed where `free` is mapped to `sea.free`...

**Review by agurfinkel — CHANGES_REQUESTED** [review]
Looks very good. A few minor things.

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:105]: Added a comment for now. This can be made less fragile but I want to solve it at the right level. 

e.g. Ideally there should be one class that deals with arranging metadata memory and all client  - MemMgr, MemToExpr classes should use its method rather than knowing internal details. We are not there yet and changing it as part of this feature appears out of scope to me.

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.hh]: done

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:83]: added comment.

**Review by priyasiddharth — COMMENTED** [review]
Please take a look

**Comment by agurfinkel** [comment]
are the build failures current @priyasiddharth ?

File: lib/seahorn/BvOpSem2Context.hh
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2Context.hh:21]: Effective Modern C++, Item 10, might be useful to review

File: test/opsem2/trackingmem/freed_unsat.c
Comment (agurfinkel) [reply test/opsem2/trackingmem/freed_unsat.c:18]: we should not map `free` to `sea.free`. They have different semantics... This also interferes with other LLVM passes that are aware of semantics of `free`. Note that there is also `delete` in C++ that should be treated similarly.

**Comment by agurfinkel** [comment]
there are still some unresolved/unanswered comments from last review

File: lib/Transforms/Scalar/PromoteMalloc.cc
Comment (agurfinkel) [inline lib/Transforms/Scalar/PromoteMalloc.cc:58]: this looks like dead code...

File: lib/seahorn/BvOpSem2Context.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2Context.hh:20]: where is the implicit cast to underlying type is used?

File: lib/seahorn/BvOpSem2ExtraWideMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2ExtraWideMemMgr.cc:645]: probably should be `WARN`

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:268]: this should be aligned load

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:71]: `std::move` got lost in conversion from `SmallVector` to `array`

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:83]: also comment what is `i == kind + 1` -- this depends on internal representation of meta-data fields.
update `for` loop per my comment -- only call `arity()` ones, use pre-inc instead of post-inc. These are my nits, but they are not hard to change...

File: tools/seahorn/seahorn.cpp
Comment (agurfinkel) [inline tools/seahorn/seahorn.cpp:320]: hmm... this might be masking the problem of mapping `free` to `sea.free`. The order of these two passes should not matter in general...

**Review by agurfinkel — CHANGES_REQUESTED** [review]
I've lost quite a bit of context, so just looking for things that stand out. 
Looks good overall. I only have minor things + build failure...

File: lib/Transforms/Scalar/PromoteMalloc.cc
Comment (priyasiddharth) [reply lib/Transforms/Scalar/PromoteMalloc.cc:58]: removed change... this was for debugging.

File: lib/seahorn/BvOpSem2Context.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2Context.hh:20]: When changing MemValTy to Expression and vice versa

File: tools/seahorn/seahorn.cpp
Comment (priyasiddharth) [reply tools/seahorn/seahorn.cpp:320]: removed change... was debugging and forgot to remove.

File: test/opsem2/trackingmem/freed_unsat.c
Comment (agurfinkel) [reply test/opsem2/trackingmem/freed_unsat.c:18]: The semantics of `sea.free()` is to mark a memory region as freed. The memory region itself, is not freed until the call to `free()`. The semantics of `free()` is to release the address space to be allocated again.

We can, for example, provide an implementation of `free()` that first calls `sea.free()` to mark memory as freed, and then calls the builtin `free()`.  We do similar things for functions like `memcmp` that have builtin versions.

The reason why we cannot change `free()` to any other name is that both `malloc()` and `free()` are understood by LLVM. That is, they are builtin functions with well defined semantics.

To me, `sea.free` is more like `llvm.lifetime.end`, if that analogy helps.

**Comment by agurfinkel** [comment]
build still fails. some prior comments are still not resolved. semantics of `sea.free` might need face-to-face discussion.

**Review by priyasiddharth — COMMENTED** [review]
Build failure may be due to https://github.com/seahorn/sea-dsa/pull/111 .. can only verify after that is checked-in.

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.cc
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:281]: please add comment in the code for that. Probably should expose this functionality eventually, but I'm ok with the current solution

Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.cc:281]: Done

**Comment by agurfinkel** [comment]
los track. Am I the bottleneck? I see CI is failing, maybe it needs to be re-triggered?

File: test/opsem2/trackingmem/freed_unsat.c
Comment (priyasiddharth) [reply test/opsem2/trackingmem/freed_unsat.c:18]: If changing `free` to `sea.free` is blocking this change then we need to discuss this face to face. Otherwise, I can submit it. The cause of build failure is well understood and will be fixed with next commit. I want to send a final cleanup commit once the above discussion is concluded.

Comment (priyasiddharth) [reply test/opsem2/trackingmem/freed_unsat.c]: Added a `sea_free` user facing function which maps to `sea.free` builtin.

**Comment by priyasiddharth** [comment]
This is ready for review.

I can't find an easy way to look at open review comments in a single list. If I missed any, please highlight if possible.

Once this is approved. I will rebase, squash the commits and submit to `dev10`.

**Review by agurfinkel — APPROVED** [review]

**Pull request merged**
